File: solutions/python/robot_battery.py
# Robot_battery.cpp 848813e190b1b85a8e75107fe8513c3be38ad1a9
import sys
import random
import itertools
import operator


# @include
def find_battery_capacity(h):
    min_height = h[0]
    capacity = 0
    for height in h:
        capacity = max(capacity, height - min_height)
        min_height = min(min_height, height)
    return capacity
# @exclude


def check_ans(h):
    """O(n^2) checking answer."""
    cap = 0
    for i in range(1, len(h)):
        for j in range(i):
            cap = max(cap, h[i] - h[j])
    return cap


def main():
    for _ in range(1000):
        if len(sys.argv) == 2:
            n = int(sys.argv[1])
        else:
            n = random.randint(1, 1000)
        a = [random.randint(0, sys.maxsize) for i in range(n)]
        print(find_battery_capacity(a))
        assert check_ans(a) == find_battery_capacity(a)


if __name__ == '__main__':
    main()

File: solutions/cpp/Robot_battery.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <limits>
#include <random>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::max;
using std::min;
using std::numeric_limits;
using std::random_device;
using std::uniform_int_distribution;
using std::vector;

// @include
int FindBatteryCapacity(const vector<int>& h) {
  int min_height = numeric_limits<int>::max(), capacity = 0;
  for (const int &height : h) {
    capacity = max(capacity, height - min_height);
    min_height = min(min_height, height);
  }
  return capacity;
}
// @exclude

// O(n^2) checking answer.
int CheckAns(const vector<int>& h) {
  int cap = 0;
  for (int i = 1; i < h.size(); ++i) {
    for (int j = 0; j < i; ++j) {
      cap = max(cap, h[i] - h[j]);
    }
  }
  return cap;
}

int main(int argc, char *argv[]) {
  default_random_engine gen((random_device())());
  for (int times = 0; times < 1000; ++times) {
    int n;
    if (argc == 2) {
      n = atoi(argv[1]);
    } else {
      uniform_int_distribution<int> dis(1, 10000);
      n = dis(gen);
    }
    vector<int> A;
    uniform_int_distribution<int> dis(0, numeric_limits<int>::max());
    for (int i = 0; i < n; ++i) {
      A.emplace_back(dis(gen));
    }
    cout << FindBatteryCapacity(A) << endl;
    assert(CheckAns(A) == FindBatteryCapacity(A));
  }
  return 0;
}
